### DISCUSSION TITLE
RabbitTemplate with SIMPLE correlation causes uncached channels

### DISCUSSION THREAD

**Discussion by bjoernhaeuser** [spring-projects/spring-amqp#2571] [Q&A]
Hello everyone,

I hope this discussion is at the right place. 

We are using a CachingConnectionFactory with `CachingConnectionFactory.ConfirmType.SIMPLE` as publisher confirm type setting.

Then we are using the RabbitTemplate something like this:
```
this.amqpTemplate.invoke(operations -> {
    // send messages

    // operations.waitForConfirmsOrDie(10_000);
}   
```

I thought this reuses channels, but after looking at the RabbitMQ stats I see a relatively high channel churn (in our case 50-100 channels/sec) and digging into this it looks like the RabbitTemplate is creating a new channel for every `.invoke()` call. 

After a debugging session through the code I saw that the `RabbitTemplate::invoke` method calls `RabbitUtils.setPhysicalCloseRequired(channel, true);` [Reference](https://github.com/spring-projects/spring-amqp/blob/v3.1.0/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/RabbitTemplate.java#L2323), when the connectionFactory says that publisher confirms are deactivated. 

Right now the `connectionFactory.isPublisherConfirms()` would only return true when the `ConfirmType` is set to `CORRELATED` [Reference](https://github.com/spring-projects/spring-amqp/blob/v3.1.0/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java#L388).

I am not familiar enough to judge if this is the right condition, but it looks to be that it should be changed to say `return !ConfirmType.NONE.equals(this.confirmType)`.

Changing the value to be of confirmType `CORRELATED` changes the behaviour and channels are properly cached. Can someone please advise if my findings are correct and/or if I am using it simply wrong?

Thank you very much!

**Comment by artembilan** [comment]
The `SIMPLE` feature was introduced here: https://github.com/spring-projects/spring-amqp/issues/2357 
However its `RabbitUtils.setPhysicalCloseRequired(channel, true);` was not guarded by the condition.
That one was fixed by this commit: https://github.com/spring-projects/spring-amqp/commit/2e99acd3335d1c0fb794eaf0340723f6944caabf

But unfortunately no any explanation in the commit message.

We can ask @garyrussell for insight, but looks like that condition in the `RabbitTemplate` must be like this:
```
				if (!connectionFactory.isPublisherConfirms() && !connectionFactory.isSimplePublisherConfirms()) {
					RabbitUtils.setPhysicalCloseRequired(channel, true);
				}
```

If we are on the same page, then I don't see why this discussion cannot be converted to issue with respective fix and so.

Thanks

**Reply by garyrussell** [reply]
That fix should work; however, I am not sure why we never cache ordinary channels used in this method.

The comment on the original commit for this method sheds some light https://github.com/spring-projects/spring-amqp/commit/7793eda

>Provide a mechanism to run `RabbitTemplate` operations in a dedicated channel
instead of returning the channel to the cache on each operation.

"...instead of..."

On reflection, I can't think of any reason to disallow reusing any channel used in this method.

I am not sure what the amqp-client does with the channel if `waitForConfirmsOrDie` fails. If it doesn't close the channel, I think we should force a physical close in that case.

**Reply by artembilan** [reply]
Thank you Gary!

The logic there is like this:
```
    public void waitForConfirmsOrDie(long timeout)
        throws IOException, InterruptedException, TimeoutException
    {
        try {
            if (!waitForConfirms(timeout)) {
                close(AMQP.REPLY_SUCCESS, "NACKS RECEIVED", true, null, false);
                throw new IOException("nacks received");
            }
        } catch (TimeoutException e) {
            close(AMQP.PRECONDITION_FAILED, "TIMEOUT WAITING FOR ACK");
            throw(e);
        }
    }
```
I'm not sure that our proxy is handling that `protected` method, so the channel is going to be physically closed.
And the next time we request it from the cache, it is going to be renewed.

Therefore it looks like we are good to reuse any channels from the cache.
Well, technically, we reuse them anyway in case of confirms:
```
channel = connection.createChannel(false);
...
if (!connectionFactory.isPublisherConfirms()) {
	RabbitUtils.setPhysicalCloseRequired(channel, true);
}
```
So, it is not clear why is that limitation was set on this `invoke()` method from day first.

I feel like this might be a behavior change for the current patch version, so we we even go a direction to fix this, it should happen in the next minor/major version.

**Reply by garyrussell** [reply]
Agreed.

**Reply by bjoernhaeuser** [reply]
First of all thank you very much for the quick response! I am not sure what are the next steps now, would you folks take it from here?  Would you expect a PR? (I am not sure what the expected changes are though).

Thank you! 🙇

**Reply by artembilan** [reply]
Created respective issue: https://github.com/spring-projects/spring-amqp/issues/2572.

But our plans is to look into that in the next major/minor version.
Probably already after holidays.

Feel free to raise a PR though!
Not promising a quick review either way 😉

**Reply by bjoernhaeuser** [reply]
Hi there!

Just wanted to revive this a bit and was looking through the discussion again. I would like to offer a PR, though I am not sure what's the result of the discussion how this should be handled. 

Should this behaviour be removed from the invoke method or should be handled differently?

Thank you! 🙇

**Reply by artembilan** [reply]
We don't have plans for major/minor version yet, so the fix (if any) is not going to be accepted yet.
You are feel free to propose some PR and we can go from there with the planning and so on.

Thanks
